Chip tool guide (#6)
markaj-nordic/connectedhomeip: docs/guides/chip_tool_guide.md

> To ensure compatibility, build the `chip-tool` and the Matter
> device from the same revision of the connectedhomeip repository.

To build and run the `chip-tool`:

1. Install all necessary packages and prepare the build system. For more
   details, see the [Building Matter](BUILDING.md) documentation:

    ```
    sudo apt-get update
    sudo apt-get upgrade

    sudo apt-get install git gcc g++ python pkg-config libssl-dev libdbus-1-dev libglib2.0-dev libavahi-client-dev ninja-build python3-venv python3-dev python3-pip unzip libgirepository1.0-dev libcairo2-dev bluez
    ```

    If the `chip-tool` is built on a Raspberry Pi, install additional
    packages and reboot the device:

    ```
    sudo apt-get install pi-bluetooth
    sudo reboot
    ```

2. Clone the Project CHIP repository:

    ```
    git clone https://github.com/project-chip/connectedhomeip.git
    ```

3. Enter the `connectedhomeip` directory:

    ```
    cd connectedhomeip
    ```

4. Initialize the git submodules:

    ```
    git submodule update --init
    ```

5. Run the following command:

    ```
    ./scripts/examples/gn_build_example.sh examples/chip-tool BUILD_PATH
    ```

>     `BUILD_PATH` specifies where the target binaries shall to be placed.

Reviewer: How about adding the additional step to change directory to BUILD_PATH? This way we could avoid pasting this BUILD_PATH text in the every invoked chip-tool command.

Author: True, I mean we can as well assume that the user runs the binary properly and stick to something like:
```
$ chip-tool
```

  | * client-generated-command-list                                                     |
  | * attribute-list                                                                    |
  | * feature-map                                                                       |
  | * cluster-revision                                                                  |
  +-------------------------------------------------------------------------------------+
[1647417857.914110][404444:404444] CHIP:TOO: Run command failure: ../../examples/chip-tool/commands/common/Commands.cpp:120: Error 0x0000002F
```
### Get the list of command options

To get the list of parameters for a specific command, run the `chip-tool` executable with the target cluster name and the target command name

Example:
```
./BUILD_PATH/chip-tool onoff on
```
Output:
```bash
[1647417976.556313][404456:404456] CHIP:TOO: InitArgs: Wrong arguments number: 0 instead of 2
Usage:
  ./chip-tool onoff on node-id/group-id endpoint-id-ignored-for-group-commands [--paa-trust-store-path] [--commissioner-name] [--trace_file] [--trace_log] [--ble-adapter] [--timedInteractionTimeoutMs] [--suppressResponse]

> [1647417976.556362][404456:404456] CHIP:TOO: Run command failure: ../../examples/chip-tool/commands/common/Commands.cpp:135: Error 0x0000002F

Reviewer: I'm not sure if I would provide this error log. It is not giving a user any useful information and code may change in time, so the documentation will be out of date then.

Author: I considered that as well and decided to leave errors from chip-tool. We can change this approach, but let's imagine that errors are not listed here and the user runs suggested commands and suddenly there are errors all over the place... that just might be confusing. We may also just mention somewhere that chip-tool reports these kind of stuff, although it's rather false positive.

Reviewer: Yeah it's kind of true. Maybe I just used to the fact that closing controller ends always with segfault, so printed errors doesn't bother me  :smile:

# Working with CHIP Tool

> The CHIP Tool (`chip-tool`) is a tool that allows to commission a Matter device

Author: @doublemis1 ?? :)

> The CHIP Tool (`chip-tool`) is a tool that allows to commission a Matter device

Reviewer 2: These two tools have the same purpose (commission and control matter device), but the platform (Android vs Linux) and way of usage are different (GUI vs Command line).

into the network and to communicate with it using the Zigbee Cluster Library

> (ZCL) messages.

Author: Although the current text actually does not lie (we are just not mentioning that apart from what is written, the tool has also other minor capabilities :) ), I will try to figure out more generic description.

But it's true that describing general application flow is rather no-go because it strictly depends on the command/sub-command/option combination.

<hr>

-   [Source files](#source)
-   [Building chip-tool](#building)
-   [Using chip-tool for Matter accessory testing](#using)
-   [Supported commands and options](#commands)

<hr>

<a name="source"></a>

## Source files

> You can find source files of the CHIP Tool in the `examples/chip-tool`

Reviewer 3: both :)

> You can find source files of the CHIP Tool in the `examples/chip-tool`

Reviewer 4: And we're ok with that? If we're to introduce this to the NCS, it will make it hard to meet the NCS requirements, yet alone to document this...

directory.

<hr>

<a name="building"></a>

## Building

Before you can use the `chip-tool`, you must compile it from the source on Linux
(amd64 / aarch64) or macOS.

> To ensure compatibility, build the `chip-tool` and the Matter device from the
> same revision of the connectedhomeip repository.

To build and run the `chip-tool`:

1. Install all necessary packages and prepare the build system. For more
   details, see the [Building Matter](BUILDING.md) documentation:

    ```
    sudo apt-get update
    sudo apt-get upgrade

    sudo apt-get install git gcc g++ python pkg-config libssl-dev libdbus-1-dev libglib2.0-dev libavahi-client-dev ninja-build python3-venv python3-dev python3-pip unzip libgirepository1.0-dev libcairo2-dev bluez
    ```

>     If the `chip-tool` is built on a Raspberry Pi, install additional packages

Author: I think the aarch64 indicates that it is supported on Raspberry Pi, but I will stress that.

    and reboot the device:

    ```
    sudo apt-get install pi-bluetooth
    sudo reboot
    ```

2. Clone the Project CHIP repository:

    ```

>     git clone https://github.com/project-chip/connectedhomeip.git

Author: Well, the 1 step  guides how to setup the build system and get all tools required to actually build the chip-tool from sources. So, IMHO checking out the sources should be kept as a separate step. Yet another approach would be to only link the [Building Matter](BUILDING.md) document as a first step - still it may introduce some conspiracy, as it also describes how to set up the Matter repository on user's PC, so we may as well squash all first 4 steps into one which only directs to the [Building Matter](BUILDING.md).

>     git clone https://github.com/project-chip/connectedhomeip.git

Reviewer 4: I'd squash to Building Matter then and add some clarifications if needed.

    ```

3. Enter the `connectedhomeip` directory:

    ```
    cd connectedhomeip
    ```

4. Initialize the git submodules:

    ```
    git submodule update --init

>     ```

Author: See my above comment regarding all first 4 steps :)

5. Run the following command:

    ```
    ./scripts/examples/gn_build_example.sh examples/chip-tool BUILD_PATH
    ```

    `BUILD_PATH` specifies where the target binaries shall to be placed.

> 6. To run the tool execute the following command:

Author: Ok, I think we can either remove the entire step 6 or just say "Check if the chip-tool has been installed /works properly" and then just ask the user to execute the first  command (`./BUILD_PATH/chip-tool`). I would opt for the second one.

    ```
    ./BUILD_PATH/chip-tool
    ```

>     As a result, a list of available clusters will show up. Each cluster can be

Author: That's totally true. The thing is that the `chip-tool` itself treats all commands as clusters in all printouts (as you pointed out). So I decided to keep the chip-tool denomination to avoid confusion, but ok, it's not really good. I will fix it.

    used as a root of the new command corresponding to ZCL data model
    interaction, for instance:

    ```
    ./BUILD_PATH/chip-tool pairing
    ```

    will print the list of sub-operations available for `pairing` cluster.

    The following is one of the possible final commands which requires
    particular command line arguments:

    ```
    ./BUILD_PATH/chip-tool pairing ble-thread
    ```

    Execution of above line will print an error and hint the proper list of
    parameters which must be provided to run ZCL operation.

<hr>

<a name="using"></a>

## Using `chip-tool` for Matter accessory testing

This section describes how to use `chip-tool` to test the Matter accessory.
Below steps depend on the application clusters that you implemented on the

> device side and may be different for your accessory.

Reviewer 4: Shouldn't we link to Configuring examples here?

Author: @greg-fer can please you point out the exact docs/link you are referring to?

### Step 1: Prepare the Matter accessory.

This tutorial is using the

> [Matter lighting-app example](https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app)

Reviewer 4: This would fit in a Requirements section.

with the Bluetooth LE commissioning. However, you can adapt this procedure to

> other available Matter examples.

Reviewer 4: Rather, the procedure is the same for other examples, no? "Adapt" suggests steps you need to take, which warrants a question "which steps" and requires documentation :)

Author: Right :) but I would like @Damian-Nordic or @doublemis1 to confirm that these steps are indeed the same for all samples (note that automatic/non-automatic BLE advertising is considered in Step 2).

Reviewer 4: Would it be possible to specify in this line that adapting means following the first 5 steps and modifying step 6?

Build and program the device with the Matter accessory firmware by following the
example's documentation.

> ### Step 2: Enable Bluetooth LE advertising on Matter accessory device.

Reviewer 4: No period in headers. Never ever. Period.

Some examples are configured to advertise automatically on boot. Other examples
require physical trigger, for example pushing a button. Follow the documentation
of the Matter accessory example to learn how Bluetooth LE advertising is enabled

> for the given example.

Reviewer 4: Well, in this case you stated we're using the lighting app example, so I would expect clear information about how the advertising is enabled for the lighting app.

Author: My doubts here applies to the comment above ,regarding the *adapt* keyword :)

    0e080000000000010000000300001335060004001fffe002084fe76e9a8b5edaf50708fde46f999f0698e20510d47f5027a414ffeebaefa92285cc84fa030f4f70656e5468726561642d653439630102e49c0410b92f8c7fbb4f9f3e08492ee3915fbd2f0c0402a0fff8
    Done
    ```

#### Wi-Fi network credentials

The following Wi-Fi network credentials must be determined to commission the
Matter accessory device to the Wi-Fi network in next steps:

-   Wi-Fi SSID
-   Wi-Fi password

Matter specification does not define how the Thread or Wi-Fi credentials are
obtained by Controller.

For example, for Thread, instead of fetching datasets directly from the Thread
Border Router, you might also use a different out-of-band method.

For Wi-Fi, the steps required to determine the SSID and password may vary
depending on the setup. For instance, you might need to contact your local Wi-Fi
network administrator.

### Step 4: Determine Matter accessory device's _discriminator_ and _setup PIN code_

The controller uses a 12-bit value called _discriminator_ to discern between
multiple commissionable device advertisements, as well as a 27-bit _setup PIN
code_ to authenticate the device. You can find these values in the logging

> terminal of the device (for instance, UART). For example:

Reviewer 4: This looks like a background information about CHIP Tool? If so, it should be described in the "application flow overview" I mentioned earlier...

Author: It's rather Matter specific. I would just replace "The controller" with  "The Matter", @greg-fer what do you think?

```
I: 254 [DL]Device Configuration:
I: 257 [DL] Serial Number: TEST_SN
I: 260 [DL] Vendor Id: 65521 (0xFFF1)
I: 263 [DL] Product Id: 32768 (0x8000)
I: 267 [DL] Hardware Version: 1
I: 270 [DL] Setup Pin Code: 20202021
I: 273 [DL] Setup Discriminator: 3840 (0xF00)
I: 278 [DL] Manufacturing Date: (not set)
I: 281 [DL] Device Type: 65535 (0xFFFF)
```

In above printout, the _discriminator_ is 3840 (0xF00) and the _setup PIN code_
is equal to 20202021 accordingly.

### Step 5: Commission Matter accessory device into existing network

#### Commissioning into Thread network over Bluetooth LE

> To commission the device to the existing Thread network run the following

Reviewer 4: But do we have a Thread network at this point? So far, we've only prepared one device and determined its credentials?

Author: The assumption is that the Thread network is up and running, so that would probably classify as a general requirement or dedicated step, but I doubt that we should describe the OTBR setup as it relies on Open Thread specific solution. Following that we would also need to describe how to setup the WiFi network.

Reviewer 4: If a Thread network is required, we need an entry under Requirements or a step that mentions setting the network up.

Author: Ok, I will add a separate step for that.

command:

```
./BUILD_PATH/chip-tool pairing ble-thread <node_id> hex:<operational_dataset> <pin_code> <discriminator>
```

where:

-   _<node_id\>_ is the user-defined ID of the node being commissioned,
-   _<operational_dataset\>_ is the Operational Dataset determined in step 3,
-   _<pin_code\>_ and _<discriminator\>_ are device specific keys determined in
    step 4

#### Commissioning into Wi-Fi network over Bluetooth LE

> 

Reviewer 4: Always introduce commands with at least a brief description.

```

The _<node_id>_ can be simply provided as a hexadecimal value with `0x` prefix.

After connecting the device over Bluetooth LE, the controller will go through
the following stages:

-   Establishing a secure connection that completes the PASE
    (Password-Authenticated Session Establishment) session using SPAKE2+
    protocol and results in printing the following log:

        ```
        Secure Session to Device Established
        ```

-   Providing the device with a network interface using ZCL Network
    Commissioning cluster commands, and the network pairing credentials from
    step 3
-   Discovering the IPv6 address of the Matter accessory using the SRP (Service
    Registration Protocol) for Thread devices, or the mDNS (Multicast Domain
    Name System) protocol for Wi-Fi or Ethernet devices. It results in printing
    log that indicates that the node address has been updated. The IPv6 address
    of the device is cached in the controller for later usage.
-   Closing the Bluetooth LE connection, as the commissioning process is
    finished and the `chip-tool` is now using only the IPv6 traffic to reach the

>     device.

Reviewer 4: This also belongs to a description of how the controller works in general. Finding it in steps is like stopping in the middle of an action sequence in a Christopher Nolan movie to discuss some science stuff. I think the user does this step and only sees the log from line 251, no?

Author: This is a copy-paste from [Python Chip Controller guide](https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/python_chip_controller_building.md) I personally don't have a strong opinion on if we should add some theory background in such guides. On the other hand maybe we should have a document describing briefly this kind of stuff, i.e. After discussion with @ArekBalysNordic we agreed to move the binding theory from light-switch-app docs here (Adding a binding table to the `binding` cluster).

Author: For now I will remove the science stuff from here :)

#### Commissioning over IP

The command below will discover devices and try to pair with the first one it
discovers using the provided _setup PIN code_.

```
./BUILD_PATH/chip-tool pairing onnetwork <node_id> <pin_code>

```

where:

-   _<node_id\>_ is the user-defined ID of the node being commissioned,
-   _<pin_code\>_ is device specific _setup PIN code_ determined in step 4

The command below will discover devices with long discriminator and try to pair
with the first one it discovers using the provided setup code.

```
./BUILD_PATH/chip-tool pairing onnetwork-long <node_id> <pin_code> <discriminator>

```

where:

-   _<node_id\>_ is the user-defined ID of the node being commissioned,
-   _<pin_code\>_ and _<discriminator\>_ are device specific keys determined in
    step 4

The command below will discover devices based on the given QR code payload
(which devices log when they start up) and try to pair with the first one it
discovers.

```
./BUILD_PATH/chip-tool qrcode <node_id> <qrcode_payload>

```

where:

-   _<node_id\>_ is the user-defined ID of the node being commissioned,

> -   _<qrcode_payload\>_ is the QR code payload

Reviewer 4: My concern for this section: are these commands I need to run one after another? Do I need to run only one depending on the scenario? If only one, mention the one, and let's create a section at the bottom of the page with "available commissioning commands" or something like this that lists all relevant commands...

Author: All commands from `Commissioning over IP` are standalone means to commission the device over IP - and the commissioning over IP and other commissioning methods (Thread/WiFi) are mutually exclusive (right, @Damian-Nordic?). Maybe I will just separate `Commissioning over IP` methods into dedicated sections and add a general description in to the main section (`Commissioning over IP`).

#### Forget the currently-commissioned device

```
./BUILD_PATH/chip-tool pairing unpair <node_id>

```

where:

-   _<node_id\>_ is the user-defined ID of the node which is going to be forgot

>     by the `chip-tool`

Reviewer 4: What is the purpose of this step?

Author: The header is kind of self-explanatory :)  But sure, I will clarify that.

#### Note

`chip-tool` currently only supports commissioning and remembering one device at
a time. The configuration state is cached in `/tmp/chip_tool_config.ini`;
deleting this and other .ini files in `/tmp` can sometimes resolve issues

> related to stale configuration.

Reviewer 4: This should go somewhere before the testing scenario, as it is a critical information about the tool.

### Step 6: Control application ZCL clusters.

For the lighting-app example, execute the following command to toggle the LED

> state:

Reviewer 4: How is this step related to the previous ones? I guess we have the device commissioned to the network, so we can start testing the clusters?

Author: Yep, that's true. Shall I add some clarification here?

Reviewer 4: Yes

```
./BUILD_PATH/chip-tool onoff toggle <node_id> <endpoint_id>
```

where:

-   _<node_id\>_ is the user-defined ID of the commissioned node
-   _<endpoint_id\>_ is the ID of the endpoint with OnOff cluster implemented

> To change the brightness of the LED, use the following command, with the

Reviewer 4: Again, is this an alternative step or something I need to do after issuing the command from line 334?

Reviewer 4: With a procedure in steps, we need to be extra clear about the order of steps and listing potential alternatives. So this entry and entry in line 321 could be a bullet point list with two alternatives, what do you think?

Author: I would not object here but for the options that are also listed as bullet points (in the whole document actually). So we would need to use different style of bullets which I am not sure is possible in Markdown (the nested bullet points are not what we want). I also worry that if I use the *alternatively* work somewhere the user may think that the commands are mutually exclusive. I will try to explain it more distinctly.

_<level\>_ equal to value between 0 and 255.

```
./BUILD_PATH/chip-tool levelcontrol move-to-level <level> <transition_time> <option_mask> <option_override> <node_id> <endpoint_id>
```

where:

-   _<level\>_ is the brightness level encoded between 0 and 255
-   _<transition_time\>_ is the transition time
-   _<option_mask\>_ is the option mask
-   _<option_override\>_ is the option override
-   _<node_id\>_ is the user-defined ID of the commissioned node
-   _<endpoint_id\>_ is the ID of the endpoint with LevelControl cluster
    implemented

In case of doubts regarding any of the options, please refer to the Matter

> specification.

Reviewer 4: Which is not yet public, so we can't advertise it like this. Even if we could, we would need to provide at least the number of the section.

### Step 7: Read basic information out of the accessory.

Every Matter accessory device supports a Basic Cluster, which maintains
collection of attributes that a controller can obtain from a device, such as the
vendor name, the product name, or software version. Use `read` command on
`basic` cluster to read those values from the device:

```
./BUILD_PATH/chip-tool basic read vendor-name <node_id> <endpoint_id>
./BUILD_PATH/chip-tool basic read product-name <node_id> <endpoint_id>
./BUILD_PATH/chip-tool basic read software-version <node_id> <endpoint_id>
```

where:

-   _<node_id\>_ is the user-defined ID of the commissioned node
-   _<endpoint_id\>_ is the ID of the endpoint with Basic cluster implemented

Use the following command to list all available commands for Basic cluster:

```
./BUILD_PATH/chip-tool basic
```

<hr>

<a name="commands"></a>

## Supported commands and options

> 

Reviewer 4: Never use headers after headers. Please put some description for this section.

  | * feature-map                                                                       |
  | * cluster-revision                                                                  |
  +-------------------------------------------------------------------------------------+
[1647417857.914110][404444:404444] CHIP:TOO: Run command failure: ../../examples/chip-tool/commands/common/Commands.cpp:120: Error 0x0000002F
```

### Get the list of command options

To get the list of parameters for a specific command, run the `chip-tool`
executable with the target cluster name and the target command name

Example:

```
./BUILD_PATH/chip-tool onoff on
```

Output:

```bash
[1647417976.556313][404456:404456] CHIP:TOO: InitArgs: Wrong arguments number: 0 instead of 2
Usage:
  ./chip-tool onoff on node-id/group-id endpoint-id-ignored-for-group-commands [--paa-trust-store-path] [--commissioner-name] [--trace_file] [--trace_log] [--ble-adapter] [--timedInteractionTimeoutMs] [--suppressResponse]
[1647417976.556362][404456:404456] CHIP:TOO: Run command failure: ../../examples/chip-tool/commands/common/Commands.cpp:135: Error 0x0000002F

```

#### Selected command options

> 

Reviewer 4: Never use headers after headers. Please add some description for this section.

```
./BUILD_PATH/chip-tool discover
```

#### Resolving node name

To resolve DNS-SD name corresponding with the given Node ID and update address
of the node in the device controller use the following command:

```
./BUILD_PATH/chip-tool discover resolve <node_id> <fabric_id>
```

where:

-   _<node_id\>_ is the ID of node to be resolved
-   _<fabric_id\>_ is the ID of the fabric where the node belongs to

#### Discovering available Matter accessory devices

To discover all Matter commissionables available in the operation area use the
following command:

```
./BUILD_PATH/chip-tool discover commissionables
```

#### Discovering available Matter commissioners

To discover all Matter commissioners available in the operation area use the
following command:

```
./BUILD_PATH/chip-tool discover commissioners
```

### Pairing

The `pairing` option supports different means regarding Matter device
commissioning procedure.

Thread and Wi-Fi commissioning use-cases were described in
[Using chip-tool for Matter accessory testing](#using) paragraph.

To list all `pairing` sub-commands use the following command:

```
./BUILD_PATH/chip-tool pairing
```

### Interacting with ZCL clusters

> As mentioned in [Using chip-tool for Matter accessory testing](#using)

Reviewer 3: Is it worth repeating that? How about just leaving examples here and keep all more generic explanations in the previous section?

2. Open a command prompt in the `connectedhomeip` directory.
3. Run the following command:

    ```
    ./scripts/examples/gn_build_example.sh examples/chip-tool BUILD_PATH
    ```

    In this command, `BUILD_PATH` specifies where the target binaries are to be
    placed.

### Running the CHIP Tool

To check if the CHIP Tool runs correctly, execute the following command from the
`BUILD_PATH` directory:

```
$ ./chip-tool
```

As a result, the CHIP Tool prints all available commands. These are called
_clusters_ in this context, but not all listed commands correspond to the
_clusters_ in the Data Model (for example, pairing or discover commands). Each
listed command can however become the root of the new more complex command by
appending it with sub-commands. Examples of specific commands and their use
cases are described in the [Supported commands and options](#commands) section.

<hr>

<a name="using"></a>

## Using CHIP Tool for Matter device testing

This section describes how to use CHIP Tool to test the Matter device. The
following steps depend on the application clusters that you implemented on the
device.

This tutorial is using the

> [Matter lighting app example](https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app)

Reviewer 4: ```suggestion
[Matter Lighting Application Example](https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app)
```

with the Bluetooth LE commissioning method support. You can use other Matter
examples and still follow this procedure. If you use a different example, the
[Step 7](#step-7-control-application-data-model-clusters) may vary depending on
the clusters implemented in your application.

### Step 1: Prepare the Matter device

Build and program the device with the Matter device firmware by following the

> [Matter lighting-app example](https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app)

Reviewer 4: ```suggestion
[Matter Lighting Application Example](https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app)
```

documentation.

### Step 2: Enable Bluetooth LE advertising on Matter device

> Some examples are configured to advertise automatically on boot. Other examples
> require physical trigger, for example pushing a button. Follow the documentation
> of the Matter device example to learn how Bluetooth LE advertising is enabled
> for the given example.

Reviewer 4: ```suggestion
Some examples are configured to advertise automatically on boot. Other examples
require physical trigger, for example pushing a button.

For the Lighting Application Example, press **Button 4** to enable the Bluetooth LE advertising.
If you are using a different example, check its documentation
for information about how to enable the Bluetooth LE advertising.
```

Author: I think we cannot assume anything about how the BLE advertising is started, the fact that the button 4 is used is specific for our platform example, but the chip-tool-guide must remain generic.

Reviewer 4: But we stopped being generic when we wrote that lighting app example is used for reference.

Author: Will be reformatted to:

"For the Lighting Application Example for the nRF Connect platform, press **Button 4** to enable the Bluetooth LE advertising."

Reviewer 3: Lighting app doesn't mean nRF connect lighting-app :) Please don't do it. We used to have battles with other developers who complained that we're trying to make "general" guides tailored to our products.

Author: OMG :)

Ahh ok, I agreed because we are already linking the Matter Lighting Application Example, but it's:
https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app

and not

https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app/nrfconnect

That being said, we indeed must remove all nrfconnect occurrences.

### Step 3: Make sure the IP network is set up

> To follow the next steps, the IP network must be up and running. For instance,
> the Thread network can be established using
> [OpenThread Border Router](https://openthread.io/codelabs/openthread-border-router#0).

Reviewer 4: ```suggestion
To follow the next steps, the IP network must be up and running. This guide assumes that you are establishing the Thread network using
[OpenThread Border Router](https://openthread.io/codelabs/openthread-border-router#0).
```

Author: I would not change that because we are actually describing the WiFi use case later on.

Reviewer 4: Yes, but in our conversation on Teams you said we're using Thread, and Wi-Fi is described as reference.

Author: Maybe I have not myself clear, sorry :) From our platform perspective the OpenThread is the use case, but others may use WiFi as well.

### Step 4: Determine network pairing credentials

You must provide the CHIP Tool with network credentials that will be used in the
device commissioning procedure to configure the device with a network interface,
such as Thread or Wi-Fi.

> The Matter specification does not define how the network credentials are to be
> obtained by controller.

Reviewer 4: ```suggestion
The Matter specification does not define the preferred way of how the network credentials are to be
obtained by controller. In this guide, we are going to obtain Thread network credentials.
```

(OTBR), retrieve this information using one of the following commands:

-   For OTBR running in Docker:

    ```
    sudo docker exec -it otbr sh -c "sudo ot-ctl dataset active -x"
    0e080000000000010000000300001335060004001fffe002084fe76e9a8b5edaf50708fde46f999f0698e20510d47f5027a414ffeebaefa92285cc84fa030f4f70656e5468726561642d653439630102e49c0410b92f8c7fbb4f9f3e08492ee3915fbd2f0c0402a0fff8
    Done
    ```

-   For OTBR native installation:

    ```
    sudo ot-ctl dataset active -x
    0e080000000000010000000300001335060004001fffe002084fe76e9a8b5edaf50708fde46f999f0698e20510d47f5027a414ffeebaefa92285cc84fa030f4f70656e5468726561642d653439630102e49c0410b92f8c7fbb4f9f3e08492ee3915fbd2f0c0402a0fff8
    Done
    ```

For Thread, you might also use a different out-of-band method to fetch the
network credentials.

#### Wi-Fi network credentials

You must get the following Wi-Fi network credentials to commission the Matter
device to the Wi-Fi network in the following steps:

-   Wi-Fi SSID
-   Wi-Fi password

The steps required to determine the SSID and password may vary depending on the
setup. For instance, you might need to contact your local Wi-Fi network
administrator.

> ### Step 5: Determine Matter device's _discriminator_ and _setup PIN code_

Reviewer 4: ```suggestion
### Step 5: Determine Matter device's discriminator and setup PIN code
```

Matter uses a 12-bit value called _discriminator_ to discern between multiple
commissionable device advertisements and a 27-bit _setup PIN code_ to
authenticate the device. You can find these values in the logging terminal of
the device (for instance, UART). For example:

```
I: 254 [DL]Device Configuration:
I: 257 [DL] Serial Number: TEST_SN
I: 260 [DL] Vendor Id: 65521 (0xFFF1)
I: 263 [DL] Product Id: 32768 (0x8000)
I: 267 [DL] Hardware Version: 1
I: 270 [DL] Setup Pin Code: 20202021
I: 273 [DL] Setup Discriminator: 3840 (0xF00)
I: 278 [DL] Manufacturing Date: (not set)
I: 281 [DL] Device Type: 65535 (0xFFFF)
```

> 
> In above printout, the _discriminator_ is `3840 (0xF00)` and the _setup PIN
> code_ is equal to `20202021`.

Reviewer 4: ```suggestion

In this snippet, the discriminator is ``3840 (0xF00)`` and the setup PIN
code is equal to ``20202021``.
These values are automatically printed in the logging terminal for each device when the device boots up.
```

Author: But it's already mentioned above:

`You can find these values in the logging terminal of
the device (for instance, UART). For example:`

Reviewer 4: OK, can you edit to something like:
`You can find these values in the logging terminal of the device (for instance, UART) when the device boots up. For example:`

### Step 6: Commission Matter device into existing IP network

> Before communicating with the Matter device, first have it join the existing IP
> network. Matter devices may use different commissioning channel. Typically,
> devices which are not yet connected to the target IP network use Bluetooth LE as
> the commissioning channel. On the other hand, if the device is already joined to
> the IP network it is enough to only commission it to the Matter network over IP
> protocol.

Reviewer 4: ```suggestion
Before communicating with the Matter device, first it must join an existing IP
network.

Matter devices may use different commissioning channel. Typically,
devices which are not yet connected to the target IP network use Bluetooth LE as
the commissioning channel. However, if the device has already joined an
IP network, the only thing needs is to commission it to the Matter network over the IP
protocol.
```

```
$ ./chip-tool pairing ble-thread <node_id> hex:<operational_dataset> <pin_code> <discriminator>
```

In this command:

-   _<node_id\>_ is the user-defined ID of the node being commissioned.
-   _<operational_dataset\>_ is the Operational Dataset determined in the
    step 3.
-   _<pin_code\>_ and _<discriminator\>_ are device-specific keys determined in
    the step 4.

##### Commissioning into Wi-Fi network over Bluetooth LE

To commission the device to the existing Wi-Fi network, use the following
command pattern:

```
$ ./chip-tool pairing ble-wifi <node_id> <ssid> <password> <pin_code> <discriminator>
```

In this command:

-   _<node_id\>_ is the user-defined ID of the node being commissioned.
-   _<ssid\>_ and _<password\>_ are credentials determined in the step 3.
-   _<pin_code\>_ and _<discriminator\>_ are device-specific keys determined in
    the step 4.

If you prefer the hexadecimal format, use the `hex:` prefix. For example:

```
$ ./chip-tool pairing ble-wifi <node_id> hex:<ssid> hex:<password> <pin_code> <discriminator>
```

> **Note:** The _<node_id>_ can be provided as a hexadecimal value with the `0x`
> prefix.

#### Commissioning into a network over IP

> This option is available when the Matter device is already present in an IP
> network, but it has not been commissioned to a Matter network yet.

Reviewer 4: ```suggestion
This option is available when the Matter device has already joined an IP
network, but it has not been commissioned to a Matter network yet.

To commission the device, you can use either the setup PIN code or the discriminator, both of which you obtained in the step 5. Alternatively, you can also use a QR code payload.
```

Author: It's rather PIN code **or** PIN code **and** discriminator.

Reviewer 4: OK, can you make edits based on this?

Author: Done already

-   _<pin_code\>_ and _<discriminator\>_ are device specific keys determined in
    the step 4.

##### Commissioning with QR code payload

Matter devices log the QR code and its payload when they boot.

Use the following command pattern to discover devices based on the given QR code
payload and try to pair with the first discovered one:

```
$ ./chip-tool qrcode <node_id> <qrcode_payload>
```

In this command:

-   _<node_id\>_ is the user-defined ID of the node being commissioned.
-   _<qrcode_payload\>_ is the QR code payload ID, for example
    `MT:Y.K9042C00KA0648G00`.

#### Forgetting the already-commissioned device

The following command removes the device with the given node ID from the list of
commissioned Matter devices:

```
$ ./chip-tool pairing unpair <node_id>
```

In this command, _<node_id\>_ is the user-defined ID of the node which is going
to be forgotten by the CHIP Tool.

### Step 7: Control application Data Model clusters

Having completed all previous steps, you have the Matter device successfully
commissioned to the network. You can now test the device by interacting with
Data Model clusters.

> For instance, in case of the lighting app example, the application has `onoff`
> and `levelcontrol` clusters implemented. This means that you can test by
> toggling the bulb (using `onoff`) or manipulating its brightness (using
> `levelcontrol`).

Reviewer 4: ```suggestion
For instance, in case of the lighting application, the application has the On/Off
and Level Control clusters implemented. This means that you can test it by
toggling the bulb (using the ``onoff`` cluster commands) or manipulating its brightness (using
the ``levelcontrol`` cluster commands).
```

Use the following command pattern to toggle the LED state:

```
$ ./chip-tool onoff toggle <node_id> <endpoint_id>
```

In this command:

-   _<node_id\>_ is the user-defined ID of the commissioned node.
-   _<endpoint_id\>_ is the ID of the endpoint with OnOff cluster implemented.

Alternatively, use the following command pattern to change the brightness of the
LED:

```
$ ./chip-tool levelcontrol move-to-level <level> <transition_time> <option_mask> <option_override> <node_id> <endpoint_id>
```

In this command:

-   _<level\>_ is the brightness level encoded between `0` and `254`, unless a
    custom range is configured in the cluster.
-   _<transition_time\>_ is the transition time.
-   _<option_mask\>_ is the option mask.
-   _<option_override\>_ is the option override.
-   _<node_id\>_ is the user-defined ID of the commissioned node.
-   _<endpoint_id\>_ is the ID of the endpoint with LevelControl cluster
    implemented.

### Step 8: Read basic information from the Matter device

> Every Matter device supports a Basic Cluster, which maintains the collection of

Reviewer 4: ```suggestion
Every Matter device supports the Basic cluster, which maintains the collection of
```

# Working with the CHIP Tool

The CHIP Tool (`chip-tool`) is a Matter controller implementation that allows to
commission a Matter device into the network and to communicate with it using
Matter messages, which may encode Data Model actions, such as cluster commands.
The tool also provides other utilities specific to Matter, such as parsing of
the setup payload or performing discovery actions.

<hr>

-   [Source files](#source)
-   [Building and running the CHIP Tool](#building)
-   [Using the CHIP Tool for Matter device testing](#using)
-   [Supported commands and options](#commands)

<hr>

<a name="source"></a>

## Source files

You can find source files of the CHIP Tool in the `examples/chip-tool`
directory.

> > **Note:** The CHIP Tool currently only supports commissioning and remembering

Reviewer 3: Not sure what it means. You can commission and remember multiple devices.

> > **Note:** The CHIP Tool currently only supports commissioning and remembering

Reviewer 2: I think it is not true. Chip tool is able to commission multiple devices and control them but not in parallel. @Damian-Nordic do you agree with me?

> one device at a time. The configuration state is cached in the
> `/tmp/chip_tool_config.ini` file. Deleting this and other `.ini` files in the
> `/tmp` directory can sometimes resolve issues related to stale configuration.

<hr>

<a name="building"></a>

## Building and running the CHIP Tool

Before you can use the CHIP Tool, you must compile it from source on Linux

> (amd64/aarch64) or macOS. If you want to build on Raspberry Pi, it must run

Reviewer 2: I will slightly change the Rasp's OS requirement. The OS must be Ubuntu but the recommended version is 20.04

> (amd64/aarch64) or macOS. If you want to build on Raspberry Pi, it must run
> Ubuntu 20.04.

Reviewer 3: ```suggestion
(amd64/aarch64) or macOS. If you want to run it on Raspberry Pi, it must use
a 64-bit OS.
```
Other Ubuntu versions should be just fine.

> **Note:** To ensure compatibility, always build the CHIP Tool and the Matter
> device from the same revision of the `connectedhomeip` repository.

### Building the CHIP Tool

To build and run the CHIP Tool:

1. Install all required packages for Matter and prepare the source code and the
   build system. Read the [Building Matter](BUILDING.md) guide for instructions.
2. Open a command prompt in the `connectedhomeip` directory.
3. Run the following command:

    ```
    ./scripts/examples/gn_build_example.sh examples/chip-tool BUILD_PATH
    ```

    In this command, `BUILD_PATH` specifies where the target binaries are to be
    placed.

### Running the CHIP Tool

To check if the CHIP Tool runs correctly, execute the following command from the
`BUILD_PATH` directory:

```

> $ chip-tool

Reviewer 3: ```suggestion
$ ./chip-tool
```
Same in other places

```

As a result, the CHIP Tool prints all available commands. These are called
_clusters_ in this context, but not all listed commands correspond to the
_clusters_ in the Data Model (for example, pairing or discover commands). Each
listed command can however become the root of the new more complex command by
appending it with sub-commands. Examples of specific commands and their use
cases are described in the [Supported commands and options](#commands) section.

<hr>

<a name="using"></a>

## Using CHIP Tool for Matter device testing

This section describes how to use CHIP Tool to test the Matter device. The
following steps depend on the application clusters that you implemented on the
device.

This tutorial is using the
[Matter lighting app example](https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app)

> with the Bluetooth LE commissioning. You can use other Matter examples and still

Reviewer 3: You also seem to cover "on-network" commissioning so BLE is not really an assumption, right?

Author: It's rather concerning the app itself. The following sentence aims to explain that it's a generic solution. I will try to clarify that.

Reviewer 3: Ok, but the app can be run without BLE, too. Anyway, feel free to disregard that comment.

such as Thread or Wi-Fi.

The Matter specification does not define how the Thread or Wi-Fi credentials are
to be obtained by controller.

#### Thread network credentials

Fetch and store the current Active Operational Dataset from the Thread Border
Router. This step may vary depending on the Thread Border Router implementation.

If you are using
[OpenThread Border Router](https://openthread.io/codelabs/openthread-border-router#0)
(OTBR), retrieve this information using one of the following commands:

-   For OTBR running in Docker:

    ```
    sudo docker exec -it otbr sh -c "sudo ot-ctl dataset active -x"
    0e080000000000010000000300001335060004001fffe002084fe76e9a8b5edaf50708fde46f999f0698e20510d47f5027a414ffeebaefa92285cc84fa030f4f70656e5468726561642d653439630102e49c0410b92f8c7fbb4f9f3e08492ee3915fbd2f0c0402a0fff8
    Done
    ```

-   For OTBR native installation:

    ```
    sudo ot-ctl dataset active -x
    0e080000000000010000000300001335060004001fffe002084fe76e9a8b5edaf50708fde46f999f0698e20510d47f5027a414ffeebaefa92285cc84fa030f4f70656e5468726561642d653439630102e49c0410b92f8c7fbb4f9f3e08492ee3915fbd2f0c0402a0fff8
    Done
    ```

> For Thread, you might also use a different out-of-band method to fetch the

Author: @Damian-Nordic, @doublemis1 is this clear enough?

Reviewer 3: Fine by me :)

device to the Wi-Fi network in the following steps:

-   Wi-Fi SSID
-   Wi-Fi password

The steps required to determine the SSID and password may vary depending on the
setup. For instance, you might need to contact your local Wi-Fi network
administrator.

### Step 5: Determine Matter device's _discriminator_ and _setup PIN code_

Matter uses a 12-bit value called _discriminator_ to discern between multiple
commissionable device advertisements and a 27-bit _setup PIN code_ to
authenticate the device. You can find these values in the logging terminal of
the device (for instance, UART). For example:

```
I: 254 [DL]Device Configuration:
I: 257 [DL] Serial Number: TEST_SN
I: 260 [DL] Vendor Id: 65521 (0xFFF1)
I: 263 [DL] Product Id: 32768 (0x8000)
I: 267 [DL] Hardware Version: 1
I: 270 [DL] Setup Pin Code: 20202021
I: 273 [DL] Setup Discriminator: 3840 (0xF00)
I: 278 [DL] Manufacturing Date: (not set)
I: 281 [DL] Device Type: 65535 (0xFFFF)
```

In above printout, the _discriminator_ is `3840 (0xF00)` and the _setup PIN
code_ is equal to `20202021`.

### Step 6: Commission Matter device into existing network

Before communicating with the Matter device, first have it join the existing

> Thread or Wi-Fi network. Depending on your network setup, this can be done over

Author: Not sure if we can say it depends on the network, rather on the Matter device being commissioned to Matter network or just joined to the WiFi/Thread net. @Damian-Nordic, @doublemis1 any opinions on that?

Reviewer 2: The device doesn't need to join the existing network before commissioning. to communicate with chip-tool. Maybe a better word will be controlling. During commissioning, the chip tool communicates with the Matter device.

Regarding dependency, in current chip tool implementation, the commissioning transport depends more on the state of accessory, BLE if the device is not in the network, and IP if the device is in Thread or WiFi network but not in Matter network.

Reviewer 3: I would simply say sth like "Matter devices may use different commissioning channel. Typically, devices which are not yet connected to the target IP network use Bluetooth LE as the commissioning channel."

Author: I think @greg-fer wanted to explicitly mention here all possible scenarios. Will the following fit?

"Matter devices may use different commissioning channel.
Typically, devices which are not yet connected to the target IP network
use Bluetooth LE as the commissioning channel. On the other hand, if the device
is already joined to the Thread or Wi-Fi network it is enough to only
commission it the Matter network with IP protocol."

Reviewer 3: Thus we need to also mention Ethernet.

> prefix.

#### Commissioning into a network over IP

This option is available when the Matter device is already present in an IP
network (Thread or Wi-Fi enabled), but it has not been commissioned to a Matter
network yet.

##### Commissioning with setup PIN code

Use the following command pattern to discover devices and try to pair with the
first discovered one using the provided setup code:

```
$ chip-tool pairing onnetwork <node_id> <pin_code>
```

In this command:

-   _<node_id\>_ is the user-defined ID of the node being commissioned.
-   _<pin_code\>_ is device specific _setup PIN code_ determined in the step 4
    and used to discover the device.

##### Commissioning with long discriminator

Use the following command pattern to discover devices with a long discriminator
and try to pair with the first discovered one using the provided setup code.

```
$ chip-tool pairing onnetwork-long <node_id> <pin_code> <discriminator>
```

In this command:

-   _<node_id\>_ is the user-defined ID of the node being commissioned.
-   _<pin_code\>_ and _<discriminator\>_ are device specific keys determined in
    the step 4.

##### Commissioning with QR code payload

> Matter devices log the QR code and its payload when they boot.

Author: Technically speaking, the log does not contain the QR code, but a proper link to a site with QR code and its payload. @Damian-Nordic, @doublemis1? BTW. Is it obligatory for all Matter devices? I guess no, since it would mandate the GUI for all accessories. Maybe better just write that the Matter device must somehow inform the user about the credentials, e.g. the QR code might be as well printed on the device's casing.

Reviewer 2: You are right, the manufacturer needs somehow provide information or a way to obtain onboarding parameters:
generated QR code based on base-38 payload
emulated NFC with base-38 payload
generated manual pairing code
In our examples, we provide information through logs (it can be also achieved by NFC with proper config, but it doesn;t work with chip tool)

Use the following command pattern to discover devices based on the given QR code
payload and try to pair with the first discovered one:

```
$ chip-tool qrcode <node_id> <qrcode_payload>
```

In this command:

-   _<node_id\>_ is the user-defined ID of the node being commissioned.
-   _<qrcode_payload\>_ is the QR code payload ID, for example
    `MT:Y.K9042C00KA0648G00`.

> #### Forgetting the currently-commissioned device

Reviewer 3: ```suggestion
#### Forgetting an already-commissioned device
```

commissioned Matter devices:

```
$ chip-tool pairing unpair <node_id>
```

In this command, _<node_id\>_ is the user-defined ID of the node which is going
to be forgotten by the CHIP Tool.

### Step 7: Control application Data Model clusters

Having completed all previous steps, you have the Matter device successfully
commissioned to the network. You can now test the device by interacting with
Data Model clusters.

For instance, in case of the lighting app example, the application has `onoff`
and `levelcontrol` clusters implemented. This means that you can test by
toggling the bulb (using `onoff`) or manipulating its brightness (using
`levelcontrol`).

Use the following command pattern to toggle the LED state:

```
$ chip-tool onoff toggle <node_id> <endpoint_id>
```

In this command:

-   _<node_id\>_ is the user-defined ID of the commissioned node.
-   _<endpoint_id\>_ is the ID of the endpoint with OnOff cluster implemented.

Alternatively, use the following command pattern to change the brightness of the
LED:

```
$ chip-tool levelcontrol move-to-level <level> <transition_time> <option_mask> <option_override> <node_id> <endpoint_id>
```

In this command:

> -   _<level\>_ is the brightness level encoded between `0` and `255`.

Reviewer 3: ```suggestion
-   _<level\>_ is the brightness level encoded between `0` and `254`, unless a custom range is configured in the cluster.
```

  +-------------------------------------------------------------------------------------+
[1647502596.396299][411686:411686] CHIP:TOO: Run command failure: ../../examples/chip-tool/commands/common/Commands.cpp:84: Error 0x0000002F
```

According to this list, the `binding` cluster supports operations such as read
or write. Attributes from that cluster can also be subscribed by the controller,
which means that the CHIP Tool will receive notifications, for instance when the
attribute value is changed or a specific event happens.

#### Examples

This section lists examples of the CHIP Tool commands dedicated to specific use
cases.

##### Writing ACL to the `accesscontrol` cluster

The Access Control List (ACL) concept allows to govern all Data Model
interactions (such as read attribute, write attribute, invoke command). For more
information about ACL, see
[Access Control Guide](https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/access-control-guide.md).

To write ACL to the `accesscontrol` cluster, use the following command pattern:

```
$ chip-tool accesscontrol write acl <acl_data> <node_id> <endpoint_id>
```

In this command:

> -   _<acl_data\>_ is the ACL data formatted as a JSON array.

Reviewer 3: Since it's a rather complex structure, some example would be pretty useful.

Author: The assumption was that the concrete commands shall be a part of guides for exemplary applications. But I have no strong opinion here, @greg-fer what do you think?

Reviewer 4: I think an example is most welcome.

-   _<node_id\>_ is the ID of the node that is going to receive ACL.
-   _<endpoint_id\>_ is the ID of the endpoint on which the `accesscontrol`
    cluster is implemented.

##### Adding a binding table to the `binding` cluster

Binding describes a relationship between the device that contains the binding
cluster and the end device. The proper ACL must be added to allow the end device
to receive commands from the bonded device. After the binding process, the
bonded device contains information about connected device, such as IPv6 address
and the route to the endpoint in the Matter network. This allows to send command

> directly from the bonded device to the end device without using the border

Reviewer 3: BR is not necessary. The communication happens just like in any other case, so it's not worth adding more constraints in this description.

controller, which means that the `chip-tool` will receive notifications, for
instance: in case the attribute value is changed or a particular event happens.

#### Examples

This is a list of `chip-tool` exemplary commands dedicated to particular use cases.

##### Writing ACL (Access Control List) to the `accesscontrol` cluster

> More information regarding ACL can be found in the [Access Control Guide](https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/access-control-guide.md).

Reviewer 5: What about moving this sentence below an example? You wrote "More information..." but there is no previous information on this topic. Or maybe simply add one sentence more at the beginning.
